src/decode.h

```diff
@@ -241,6 +241,18 @@ typedef struct {
 
     /* instruction length */
     uint8_t insn_len;
+    /* According to Tail-call optimization (TCO), if a C function ends with
+     * a function call to another function or itself and simply returns that
+     * function's result, the compiler can substitute a simple jump to the
+     * other function for the 'call' and 'return' instructions . The self
+     * -recursive function can therefore use the same function stack frame.
+     *
+     * Member tailcall help us to determine whether the IR is last IR in
+     * a basic block. We can use this member to rewrite the function emulate
+     * into a self-recursive version, allowing the compiler to do TCO.
+     */
+    bool tailcall;
+    void *opfunc;
```

Comment:
Declare `opfunc` with typedef.

Reply:
The type of `opfunc` is `(*func)(riscv_t *rv, rv_insn_t *ir)`, it need the definition of `rv_insn_t`, but `opfunc` is also defined in rv_insn_t. That's the reason why I choose void type here.

Reply:
Do `#include "riscv.h"` and apply the change below:
```diff
@@ -226,7 +228,7 @@ enum {
     INSN_32 = 4,
 };
 
-typedef struct {
+typedef struct rv_insn {
     union {
         int32_t imm;
         uint8_t rs3;
@@ -252,7 +254,7 @@ typedef struct {
      * into a self-recursive version, allowing the compiler to do TCO.
      */
     bool tailcall;
-    void *opfunc;
+    bool (*opfunc)(riscv_t *, const struct rv_insn *);
 } rv_insn_t;
 
 /* decode the RISC-V instruction */
```

Reply:
Alternatively, you can combine `tailcall` and `opfunc` into a single member in `rv_insn_t`. See `src/map.h` and commit 7fb4ff754e48f09e21f0e21137fd28ce59361de5 for details.